core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageInputStreamTest.java

```diff
@@ -318,4 +319,119 @@ void toInputStream_executor() throws IOException {
+                                             .stream()
+                                             .map(x -> x.toString().getBytes())
+                                             .reduce(Bytes::concat).get();
+
+        final ByteBuf result = Unpooled.buffer();
+        int read;
+        while ((read = inputStream.read()) != -1) {
+            result.writeByte(read);
+        }
+
+        final int readableBytes = result.readableBytes();
+        final byte[] actual = new byte[readableBytes];
+        for (int i = 0; i < readableBytes; i++) {
+            actual[i] = result.readByte();
+        }
+        result.release();
+        assertThat(actual).isEqualTo(expected);
+        assertThat(inputStream.available()).isZero();
+    }
+
+    @Test
+    void maybeRequest_when_not_enough_data() throws IOException {
+        final StreamMessage<String> streamMessage1 = StreamMessage.of("12", "34");
+        final DefaultStreamMessage<String> streamMessage2 = new DefaultStreamMessage<>();
+        streamMessage2.write("56");
+        streamMessage2.write("78");
+        final AtomicBoolean consumed = new AtomicBoolean();
+        streamMessage2.whenConsumed().thenRun(() -> {
+            consumed.getAndSet(true);
+            streamMessage2.close();
+        });
+        final InputStream inputStream = StreamMessage
+                .concat(streamMessage1, streamMessage2)
+                .toInputStream(x -> HttpData.wrap(x.getBytes()));
+
+        final ByteBuf result = Unpooled.buffer();
+        for (int i = 0; i < 2; i++) {
+            result.writeByte(inputStream.read());
+        }
+
+        int readableBytes = result.readableBytes();
+        byte[] actual = new byte[readableBytes];
+        for (int i = 0; i < readableBytes; i++) {
+            actual[i] = result.readByte();
+        }
+        assertThat(actual).isEqualTo(ImmutableList.of("12")
+                                                  .stream()
+                                                  .map(String::getBytes)
+                                                  .reduce(Bytes::concat).get());
+        assertThat(inputStream.available()).isEqualTo(0);
+        assertThat(consumed.get()).isFalse();
+
+        int read;
+        while ((read = inputStream.read()) != -1) {
+            result.writeByte(read);
+        }
+
+        readableBytes = result.readableBytes();
+        actual = new byte[readableBytes];
+        for (int i = 0; i < readableBytes; i++) {
+            actual[i] = result.readByte();
+        }
+        result.release();
+        assertThat(actual).isEqualTo(ImmutableList.of("34", "56", "78")
+                                                  .stream()
+                                                  .map(String::getBytes)
+                                                  .reduce(Bytes::concat).get());
+        assertThat(inputStream.available()).isZero();
+        assertThat(consumed.get()).isTrue();
```

Comment:
I'm not sure this guarantee a happen-before relationship. `whenComsumed()` may call in a different thread.
```suggestion
        await().untilTrue(consumed);
```

Reply:
✅ updated!